src/anndata/experimental/backed/_lazy_arrays.py

```diff
@@ -0,0 +1,155 @@
+        categories: ZarrArray | H5Array,
+        ordered: bool,
+        *args,
+        **kwargs,
+    ):
+        self._categories = categories
+        self._ordered = ordered
+        self._categories_cache = None
+        self._codes = ZarrOrHDF5Wrapper[type(codes)](codes)
+        self.shape = self._codes.shape
+        self.dtype = CategoricalDtypeAccessor(
+            categories=self._categories, ordered=self._ordered
+        )
+
+    @property
+    def categories(self):  # __slots__ and cached_property are incompatible
+        if self._categories_cache is None:
+            if isinstance(self._categories, ZarrArray):
+                self._categories_cache = self._categories[...]
+            else:
+                if (
+                    "read_dataset" not in dir()
+                ):  # avoid circular dependency, not sure what caused this all of a sudden after merging https://github.com/scverse/anndata/pull/949/commits/dc9f12fcbca977841e967c8414b9f1032e069250
+                    from ..._io.h5ad import read_dataset
+                self._categories_cache = read_dataset(self._categories)
+        return self._categories_cache
+
+    def __getitem__(
+        self, key: xr.core.indexing.ExplicitIndexer
+    ) -> xr.core.extension_array.PandasExtensionArray:
+        codes = self._codes[key]
+        categorical_array = pd.Categorical.from_codes(
+            codes=codes, categories=self.categories, ordered=self._ordered
+        )
+        if settings.should_remove_unused_categories:
+            return xr.core.extension_array.PandasExtensionArray(
+                categorical_array.remove_unused_categories()
+            )
+        return xr.core.extension_array.PandasExtensionArray(categorical_array)
+
+
+class MaskedArray(BackendArray):
+    def __init__(
+        self,
+        values: ZarrArray | H5Array,
+        dtype_str: str,
+        mask: ZarrArray | H5Array | None = None,
+    ):
+        self._mask = ZarrOrHDF5Wrapper[type(mask)](mask)
+        self._values = ZarrOrHDF5Wrapper[type(values)](values)
+        self._dtype_str = dtype_str
+        self.shape = self._values.shape
+        self.dtype = pd.api.types.pandas_dtype(self._values.dtype)
+
+    def __getitem__(self, key) -> xr.core.extension_array.PandasExtensionArray:
+        values = self._values[key]
+        if self._mask is not None:
+            mask = self._mask[key]
+            if self._dtype_str == "nullable-integer":
+                # numpy does not support nan ints
+                return xr.core.extension_array.PandasExtensionArray(
+                    pd.arrays.IntegerArray(values, mask=mask)
+                )
+            elif self._dtype_str == "nullable-boolean":
+                return xr.core.extension_array.PandasExtensionArray(
+                    pd.arrays.BooleanArray(values, mask=mask)
+                )
+        return xr.core.extension_array.PandasExtensionArray(pd.array(values))
+
+
+@_subset.register(DataArray)
```

Comment:
Hm, I’m not sure if it’s good practice to register more here in another file.

I think I like it …